[CINN][New Hardware Update] SYCL first PR

feixi21 opened the pull request:
<!-- TemplateReference: https://github.com/PaddlePaddle/Paddle/wiki/PULL-REQUEST-TEMPLATE--REFERENCE -->
<!-- Demo: https://github.com/PaddlePaddle/Paddle/pull/24810 -->

### PR Category
<!-- One of [ User Experience | Execute Infrastructure | Operator Mechanism | CINN | Custom Device | Performance Optimization | Distributed Strategy | Parameter Server | Communication Library | Auto Parallel | Inference | Environment Adaptation ] -->
CINN

### PR Types
<!-- One of [ New features | Bug fixes | Improvements | Performance | BC Breaking | Deprecations | Docs | Devs | Not User Facing | Security | Deprecations | Others ] -->
New features

### Description
<!-- Describe what you’ve done -->
SYCL后端第一个PR

- 添加HygonDCUArchSYCL，并实现HygonDCUArchSYCL相应处理流程

feixi21 committed:
add HygonDCU_SYCL

feixi21 committed:
DPCPP_CMAKE

feixi21 committed:
fix_merge_conflict

feixi21 committed:
add HygonDCUSYCL part two

feixi21 committed:
add sycl runtime

feixi21 committed:
fix_bugs

feixi21 committed:
fix_bugs

feixi21 committed:
fix_bugs

feixi21 committed:
fix_codestyle

feixi21 committed:
add define sycl fp16 bf16 judgment

fix_bugs

fix_bugs

ZhangX-21 commented on paddle/cinn/hlir/framework/pir/op_lowering_impl.cc:
```diff
@@ -536,6 +542,14 @@ std::vector<ir::LoweredFunc> OpLowererImpl::DoOpLower(
           } else {
             op_func_arg_tensors->push_back(expr.as_tensor_ref());
           }
+        },
```
代码实现没有差异性，std::variant<common::HygonDCUArchHIP, common::HygonDCUArchSYCL>处理 ？

feixi21 replied:
已参照修改意见修改。

ZhangX-21 commented on paddle/cinn/hlir/op/contrib/gather_nd.cc:
```diff
@@ -222,6 +222,9 @@ std::shared_ptr<framework::OpStrategy> StrategyForGatherNd(
           },
           [&](common::HygonDCUArchHIP) {
             pe::IRGpuScheduleInjective(ir_sch, output_shapes.front(), target);
+          },
```
同上？

feixi21 replied:
已参照修改意见修改。

ZhangX-21 commented on paddle/cinn/hlir/op/contrib/repeat.cc:
```diff
@@ -224,6 +224,9 @@ std::shared_ptr<framework::OpStrategy> StrategyForRepeat(
           },
           [&](common::HygonDCUArchHIP) {
             pe::IRGpuScheduleInjective(ir_sch, output_shapes.front(), target);
+          },
```
同上？

feixi21 replied:
已参照修改意见修改。

ZhangX-21 commented on paddle/cinn/runtime/sycl/sycl_backend_api.cc:
```diff
@@ -0,0 +1,281 @@
+                                ::sycl::info::device_type::gpu)[0]
+                                .get_backend());
+      },
+      [&](common::X86Arch) { LOG(FATAL) << "SYCL Not supported this arch \n"; },
```
同上, arch 为sycl /hip 的backend为都是:sycl::backend::ext_oneapi_hip，不会歧义吗？

feixi21 replied:
已参照修改意见修改。

ZhangX-21 commented on paddle/cinn/runtime/sycl/sycl_backend_api.cc:
```diff
@@ -0,0 +1,281 @@
+  this->queues.resize(this->devices.size());
+  // sycl::backend -> Target::Arch
+  switch (backend) {
+    case ::sycl::backend::ext_oneapi_hip:
```
在更早的时候是不是就拦截掉了？

feixi21 replied:
之前函数返回SYCLBackendAPI类内的私有变量arch，因此之前此处需要根据backend初始化this->arch，已参照修改意见修改。

ZhangX-21 commented on paddle/cinn/runtime/sycl/sycl_backend_api.cc:
```diff
@@ -0,0 +1,281 @@
+              .get_info<::sycl::info::device::sub_group_sizes>();
+      size_t max_sub_group_size =
+          *max_element(std::begin(sub_group_sizes), std::end(sub_group_sizes));
+      rv = static_cast<int>(max_sub_group_size);
```
缺少break?

feixi21 replied:
已参照修改意见修改。

ZhangX-21 commented on paddle/cinn/runtime/sycl/sycl_module.cc:
```diff
@@ -0,0 +1,62 @@
+SYCLModule::~SYCLModule() {
+  VLOG(3) << "destructor for SYCLModule";
+  if (so_handler_ != nullptr) {
+    // dlclose(so_handler_);
```
冗余代码？

feixi21 replied:
已参照修改意见修改。

feixi21 committed:
modify according to comment

modify according to comment

modify according to comment

modify according to comment

ZhangX-21 commented on paddle/cinn/backends/codegen_cuda_host.h:
```diff
@@ -60,6 +60,13 @@ class CodeGenGpuHost : public CodeGenHost {
           } else {
             return CodeGenHost::Visit(op);
           }
+        },
```
这个也是？

feixi21 replied:
HygonDCUArchHIP需要判定是否为call_hip_kernel，HygonDCUArchSYCL需要判定是否为call_sycl_kernel，所需要判定的op不同，所以这里不能合并。

ZhangX-21 commented on paddle/cinn/backends/codegen_device_util.cc:
```diff
@@ -218,6 +218,11 @@ void detail::CollectBucketStrategyHostFunctionVisitor::ProcessLoweredFunc(
       [&](common::HygonDCUArchHIP) {
 #ifdef CINN_WITH_HIP
         shared_mem_bytes = CalculateSharedMemory(func);
+#endif
```
这个也是吧

feixi21 replied:
修改了一下HygonDCUArchSYCL的对应逻辑。

ZhangX-21 commented on paddle/cinn/backends/compiler.cc:
```diff
@@ -286,6 +287,10 @@ std::string Compiler::GetSourceCode(const ir::Module& module) {
 #endif
+      },
+      [&](common::HygonDCUArchSYCL) -> std::string {
+        PADDLE_THROW(phi::errors::Unimplemented(
```
CINN_NOT_IMPLEMENTED？

feixi21 replied:
已参照修改意见修改。

ZhangX-21 commented on paddle/cinn/common/target.cc:
```diff
@@ -79,6 +86,11 @@ int GetRuntimeArchImpl(HygonDCUArchHIP) {
 
+int GetRuntimeArchImpl(HygonDCUArchSYCL) {
+  PADDLE_THROW(::common::errors::InvalidArgument(
+      "HygonDCUArchSYCL not supported GetRuntimeArch!"));
```
CINN_NOT_IMPLEMENTED

feixi21 replied:
已参照修改意见修改。

ZhangX-21 commented on paddle/cinn/hlir/framework/pir/op_lowering_impl.cc:
```diff
@@ -372,6 +372,12 @@ std::vector<ir::LoweredFunc> OpLowererImpl::PostProcess(
 #ifdef CINN_WITH_HIP
             optim::EliminateCommonGlobalMemoryRead(&(func_body));
             optim::OptimizeExprGPU(&(func_body));
+#endif
```
同上！

feixi21 replied:
已参照修改意见修改。

ZhangX-21 commented on paddle/cinn/runtime/sycl/sycl_util.cc:
```diff
@@ -0,0 +1,1266 @@
+    ::sycl::range<3> Grid(grid_z, grid_y, grid_x);
+    ::sycl::range<3> Block(block_z, block_y, block_x);
+    // need malloc_shared
+    // LOG(INFO) << "kernel args :" << (float* )(*(void **)(kernel_args[0]))[0]
```
冗余代码

feixi21 replied:
已参照修改意见修改。

feixi21 committed:
modify according to comment

fix_bugs

fix_bugs

ZhangX-21 commented on paddle/cinn/optim/vectorize_loops.cc:
```diff
@@ -879,15 +879,17 @@ struct VectorizeLoops_ : public IRMutator<Expr *> {
+                .Visit(&new_forloop->body);
+          },
+          [&](std::variant<common::HygonDCUArchHIP, common::HygonDCUArchSYCL>) {
+            setNvHygon();
```
setNvHygon会不会有歧义那？海光的设备为啥和NV的有关系那？

feixi21 replied:
这里是在NV设备与海光设备上均调用相同的函数，无需修改

ZhangX-21 commented on paddle/cinn/runtime/sycl/sycl_backend_api.cc:
```diff
@@ -0,0 +1,271 @@
+      [&](common::ARMArch) { LOG(FATAL) << "SYCL Not supported this arch \n"; },
+      [&](common::NVGPUArch) { backend = ::sycl::backend::ext_oneapi_cuda; },
+      [&](std::variant<common::HygonDCUArchHIP, common::HygonDCUArchSYCL>) {
+        backend = ::sycl::backend::ext_oneapi_hip;
```
common::HygonDCUArchHIP对应:hip::backend::ext_oneapi_hip?
common::HygonDCUArchSYCL对应::sycl::backend::ext_oneapi_sycl?

feixi21 replied:
已参照修改意见修改。

feixi21 committed:
fix_bugs

fix_bugs

fix_bugs

fix_bugs

feixi21 committed:
fix_merge_conflicts

feixi21 committed:
Merge branch 'develop' into paddle-sycl-PR1

phlrain commented on paddle/cinn/pybind/runtime.cc:
```diff
@@ -108,6 +108,11 @@ cinn_buffer_t *CreateBufferFromNumpyImpl(common::HygonDCUArchHIP,
+cinn_buffer_t *CreateBufferFromNumpyImpl(common::HygonDCUArchSYCL,
+                                         py::array data) {
+  PADDLE_THROW(::common::errors::Unimplemented("CINN old obsolete code!"));
+}
```
这部分代码是不是可以清理了

phlrain commented on paddle/cinn/runtime/sycl/cinn_sycl_runtime_source.h:
```diff
@@ -0,0 +1,959 @@
+
+#undef CINN_CUDA_INDEX_ADD
+
+int cinn_sycl_resize_bilinear(const int *buf,
```
这些函数是不是没有地方在调用的，是不是可以清理下

phlrain commented on paddle/cinn/runtime/sycl/sycl_util.h:
```diff
@@ -0,0 +1,219 @@
+                           int b3,
+                           int b4);
+
+void cinn_call_cnnl_conv2d_forward(void* v_args,
```
类似这种接口，是不是也不在需要了

XiaoguangHu01 reviewed (approved):
LGTM
